src/middlewared/middlewared/plugins/interface/link_address.py

```diff
@@ -0,0 +1,136 @@
+
+RE_FREEBSD_BRIDGE = re.compile(r"bridge([0-9]+)$")
+RE_FREEBSD_LAGG = re.compile(r"lagg([0-9]+)$")
+
+
+class InterfaceCollection:
+    def __init__(self, interfaces):
+        self.interfaces = interfaces
+
+    @property
+    def by_name(self):
+        return {self.get_name(i): i for i in self.interfaces}
+
+    @property
+    def by_link_address(self):
+        return {self.get_link_address(i): i for i in self.interfaces}
+
+    def __iter__(self):
+        return iter(self.interfaces)
+
+    def get_name(self, i):
+        raise NotImplementedError
+
+    def get_link_address(self, i):
+        raise NotImplementedError
+
+
+class DatabaseInterfaceCollection(InterfaceCollection):
+    def get_name(self, i):
+        return i["interface"]
+
+    def get_link_address(self, i):
+        return i["link_address"]
+
+
+class RealInterfaceCollection(InterfaceCollection):
+    def get_name(self, i):
+        return i["name"]
+
+    def get_link_address(self, i):
+        return i["state"]["link_address"]
+
+
+async def rename_interface(middleware, db_interface, name):
+    middleware.logger.info("Renaming interface %r to %r", db_interface["interface"], name)
+    await middleware.call("datastore.update", "network.interfaces", db_interface["id"],
+                          {"interface": name}, {"prefix": "int_"})
+
+    for bridge in await middleware.call("datastore.query", "network.bridge"):
+        try:
+            index = bridge["members"].index(db_interface["interface"])
+        except ValueError:
+            continue
+
+        bridge["members"][index] = name
+        middleware.logger.info("Setting bridge %r members: %r", bridge["id"], bridge["members"])
+        await middleware.call("datastore.update", "network.bridge", bridge["id"], {"members": bridge["members"]})
+
+    for lagg_member in await middleware.call("datastore.query", "network.lagginterfacemembers"):
+        if lagg_member["lagg_physnic"] == db_interface["interface"]:
+            middleware.logger.info("Setting LAGG member %r physical NIC %r", lagg_member["id"], name)
+            await middleware.call("datastore.update", "network.lagginterfacemembers", lagg_member["id"],
+                                  {"lagg_physnic": name})
+
+    for vlan in await middleware.call("datastore.query", "network.vlan"):
+        if vlan["vlan_pint"] == db_interface["interface"]:
+            middleware.logger.info("Setting VLAN %r parent NIC %r", vlan["vlan_vint"], vlan["vlan_pint"])
+            await middleware.call("datastore.update", "network.vlan", vlan["id"],
+                                  {"vlan_pint": name})
+
+
+async def setup(middleware):
+    real_interfaces = RealInterfaceCollection(await middleware.call("interface.query", [["fake", "!=", True]]))
+    db_interfaces = DatabaseInterfaceCollection(
+        await middleware.call("datastore.query", "network.interfaces", [], {"prefix": "int_"}),
+    )
+
+    # Migrate BSD network interfaces to Linux
```

Comment:
This is has to live here for eternity, until we support migrating databases from FreeBSD?
Impossible to be part of a data migration?

Reply:
It can be a part of data migration, but I've decided to put it here as this process is coupled with `# Update link addresses for interfaces in the database` which is executed on both BSD and Linux. (It is not necessary in Linux, but I've decided to keep that field populated correctly as it might have some use in the future)

Reply:
Seems reasonable, but the way it is, if any exception happens in this case middlewared startup will crash, no?
Can/should we make it more resilient?

Reply:
Fixed